File: cmd/status.go
package cmd

import (
	"dvs/internal/file"
	"dvs/internal/git"
	"dvs/internal/log"
	"dvs/internal/meta"
	"os"
	"time"

	"github.com/dustin/go-humanize"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

func runStatusCmd(cmd *cobra.Command, args []string) error {
	type jsonFileResult struct {
		Path      string `json:"path"`
		Status    string `json:"status"`
		FileSize  uint64 `json:"fileSize"`
		FileHash  string `json:"fileHash"`
		Timestamp string `json:"timestamp"`
		SavedBy   string `json:"savedBy"`
		Message   string `json:"message"`
	}

	type jsonResult struct {
		Files  []jsonFileResult `json:"files"`
		Errors []log.JsonIssue  `json:"errors"`
	}

	jsonLogger := jsonResult{}
	var metaPaths []string

	// If no arguments are provided, get the status of all files in the current git repository
	if len(args) == 0 {
		// Get git dir
		gitDir, err := git.GetNearestRepoDir(".")
		if err != nil {
			return err
		}

		metaPaths, err = meta.GetAllMetaFiles(gitDir)
		if err != nil {
			return err
		}
	} else {
		// Get meta files from globs
		metaPaths = meta.ParseGlobs(args)
	}

	// Create colors
	colorFilePulled := color.New(color.FgGreen, color.Bold).Sprint
	colorFileOutdated := color.New(color.FgHiYellow, color.Bold).Sprint
	colorFileNotPulled := color.New(color.FgRed, color.Bold).Sprint
	iconPulled := colorFilePulled("✔")
	iconOutdated := colorFileOutdated("!")
	iconNotPulled := colorFileNotPulled("✘")

	// Track number of files
	numFilesPulled := 0
	numFilesOutdated := 0
	numFilesNotPulled := 0

	// Print info about each file
	if len(metaPaths) == 0 {
		log.Print(log.ColorBold(log.ColorYellow("!")), "No files were queued")
		log.JsonLogger.Issues = append(log.JsonLogger.Issues, log.JsonIssue{
			Severity: "warning",
			Message:  "no files queued",
			Location: ".",
		})
	} else {
		log.Print(color.New(color.Bold).Sprint("file info "),
			iconPulled, "up to date ",
			iconOutdated, "out of date ",
			iconNotPulled, "not present ",
		)
	}
	for _, path := range metaPaths {
		relPath, err := git.GetRelativePath(".", path)
		if err != nil {
			log.Print(log.ColorRed("\n✘"), "Failed to get relative path", log.ColorFaint(err.Error()))
			log.JsonLogger.Issues = append(log.JsonLogger.Issues, log.JsonIssue{
				Severity: "error",
				Message:  "failed to get relative path",
				Location: path,
			})

			continue
		}

		// Get file info
		metadata, err := meta.Load(path)
		if err != nil {
			log.Print(log.ColorRed("    ✘"), "File not in devious", log.ColorFile(relPath))
			log.JsonLogger.Issues = append(log.JsonLogger.Issues, log.JsonIssue{
				Severity: "error",
				Message:  "file not in devious",
				Location: relPath,
			})

			continue
		}

		// Determine file tag based on file status
		var fileLight string
		var fileStatus string
		_, statErr := os.Stat(path)
		fileHash, hashErr := file.GetFileHash(path)
		if statErr == nil && hashErr == nil && fileHash == metadata.FileHash {
			fileLight = iconPulled
			fileStatus = "up to date"
			numFilesPulled++
		} else if statErr == nil {
			fileLight = iconOutdated
			fileStatus = "out of date"
			numFilesOutdated++
		} else {
			fileLight = iconNotPulled
			fileStatus = "not present"
			numFilesNotPulled++
		}

		// Determine whether to print timestamp
		timestamp := ""
		if !metadata.Timestamp.IsZero() {
			timestamp = metadata.Timestamp.Format(time.DateTime)
		}

		// Print file info
		log.Print("   ", fileLight,
			log.ColorFile(relPath), "",
			log.ColorFaint(humanize.Bytes(metadata.FileSize)), "",
			log.ColorFaint(timestamp), "",
			log.ColorFaint(metadata.SavedBy), "",
			log.ColorFaint(metadata.Message),
		)
		jsonLogger.Files = append(jsonLogger.Files, jsonFileResult{
			Path:      relPath,
			Status:    fileStatus,
			FileSize:  metadata.FileSize,
			FileHash:  metadata.FileHash,
			Timestamp: timestamp,
			SavedBy:   metadata.SavedBy,
			Message:   metadata.Message,
		})
	}

	// Print overview
	if len(metaPaths) > 0 {
		log.Print(color.New(color.Bold).Sprint("\ntotals"))
		log.Print(
			colorFilePulled(numFilesPulled), "up to date ",
			colorFileOutdated(numFilesOutdated), "out of date ",
			colorFileNotPulled(numFilesNotPulled), "not present ",
		)
	}

	log.Dump(jsonLogger)

	return nil
}

func getStatusCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "status [glob] [another-glob]",
		Short: "Gets the status of files tracked by devious",
		Long:  "Gets the status of files tracked by devious. If glob(s) are provided, only those globs will be checked. Otherwise, all files in the current git repository will be checked.",
		PreRun: func(cmd *cobra.Command, args []string) {
			log.PrintLogo()
		},
		RunE: runStatusCmd,
	}

	return cmd
}

File: src/library/status.rs
use std::path::PathBuf;
use anyhow::{Context, Result};
use serde::{Serialize, Deserialize};
use crate::helpers::config;
use crate::helpers::hash;
use crate::helpers::repo;
use crate::helpers::file;
use crate::helpers::parse;


#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct JsonFileResult {
    pub path: PathBuf,
    pub status: String,
    pub file_size: u64,
    pub file_hash: String,
    pub time_stamp: String,
    pub saved_by: String,
    pub message: String
}

pub fn dvs_status(files: &Vec<String>) -> Result<Vec<JsonFileResult>> {
    // struct for each file's status and such
    let mut json_logger: Vec<JsonFileResult> = Vec::new();

    // vector of files
    let mut meta_paths: Vec<PathBuf> = Vec::new();

    // if no arguments are provided, get the status of all files in the current git repository
    if files.is_empty() {
        // Get git root
        let git_dir = repo::get_nearest_repo_dir(&PathBuf::from(".")).with_context(|| "could not find git repo root - make sure you're in an active git repository")?;

        // get config
        config::read(&git_dir)?;

        // get meta files
       meta_paths = [meta_paths, parse::get_all_meta_files(git_dir)].concat();
    } // if doing all files
    else {
        meta_paths = [meta_paths, parse::parse_globs(files)].concat();
    } // else specific files

    if meta_paths.is_empty() {return Ok(json_logger)}

    json_logger  = meta_paths.into_iter().map(|path| {
        // get relative path
        let rel_path = repo::get_relative_path(&PathBuf::from("."), &path).expect("couldn't get relative path");
        
        // get file info
        let metadata = file::load(&path).expect("couldn't get metadata");
        
        // asign status: not-present by default
        let mut status = String::from("out-of-date");
        // if the file path doesn't exist assign status to "not-present"
        if !path.exists() {status = String::from("not-present")}
        else {
            // get whether file was hashable and file hash
            match hash::get_file_hash(&path) {
                Some(file_hash) => {
                    if file_hash == metadata.file_hash {
                        status = String::from("up-to-date")
                    }
                }
                None => (),
            }; 
        }
       
        // assemble info into JsonFileResult
        JsonFileResult{
            path: rel_path,
            status: status,
            file_size: metadata.file_size,
            file_hash: metadata.file_hash,
            time_stamp: metadata.time_stamp,
            saved_by: metadata.saved_by,
            message: metadata.message
        }
    }).collect::<Vec<JsonFileResult>>();

Ok(json_logger)
} // run_status_cmd